src/main/java/org/truffleruby/core/module/ModuleNodes.java

```diff
@@ -2584,4 +2584,22 @@ protected Object doClass(RubyClass rubyClass) {
             return rubyClass.isSingleton;
         }
     }
+
+    @CoreMethod(names = "undefined_instance_methods")
+    public abstract static class UndefinedInstanceMethodsNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        @TruffleBoundary
+        protected RubyArray undefinedInstanceMethods(RubyModule module) {
+            List<RubySymbol> methodNames = new ArrayList<>();
+
+            for (InternalMethod methodEntry : module.fields.getMethods()) {
+                if (methodEntry != null && methodEntry.isUndefined()) {
+                    methodNames.add(getLanguage().getSymbol(methodEntry.getName()));
+                }
+            }
+
+            return createArray(methodNames.toArray());
+        }
+    }
```

Comment:
Wondering whether it makes sense to get rid of `@TruffleBoundary` in this case.

Reply:
That would mean not using an ArrayList (and e.g. using a RubyArray instead) and Iterable is not PE-safe.
So it does not seem worth it.